### DISCUSSION TITLE
Proper way to override registration behavior

### DISCUSSION THREAD

**Discussion by gl2007** [codeigniter4/shield#823] [Q&A] [answered]
I wanted to customize registration behavior, so extended RegisterController and tried to use getValidationRules. I wanted to reuse the built-in validation logic and some other pieces.

I got "Error: Typed property CodeIgniter\Shield\Controllers\RegisterController::$tables must not be accessed before initialization".

Note that I am doing "new MyRegisterController()) and then calling ->method().

Assuming that initController has to be mandatorily overridden, I wrote this: 

```php
public function initController(
        RequestInterface $request,
        ResponseInterface $response,
        LoggerInterface $logger
    ): void {
        parent::initController(
            $request,
            $response,
            $logger
        );
}
```

but that one gave me Fatal error about some compatibility which I did not capture. 

Can you help?

**Comment by kenjis** [comment]
Can you try the following change?

```diff
--- a/src/Controllers/RegisterController.php
+++ b/src/Controllers/RegisterController.php
@@ -33,7 +33,7 @@ class RegisterController extends BaseController
     /**
      * Auth Table names
      */
-    private array $tables;
+    protected array $tables;
 
     public function initController(
         RequestInterface $request,
```

**Reply by gl2007** [reply]
I was thinking about making that protected. However, it didn't work and still gave the same error. Note that I have removed the "initController" method now.

**Comment by datamweb** [comment]
This is the way to customize the controller and I have no problem with it. Or at least I didn't get it right.

My **Routes.php** :

```php
service('auth')->routes($routes, ['except' => ['register']]);

$routes->get('register', '\App\Controllers\RegisterController::registerView');
$routes->post('register', '\App\Controllers\RegisterController::registerAction');
```
my **App\Controllers\RegisterController.php**:
```php
<?php

namespace App\Controllers;

use CodeIgniter\Shield\Controllers\RegisterController as ShieldRegisterController;

class RegisterController extends ShieldRegisterController
{
     /**
     * Displays the registration form.
     *
     * @return RedirectResponse|string
     */
    public function registerView(){
        return 'override';
    }

    public function myMethod(){
        return 'My Method!!';
    }
}
```

**Reply by gl2007** [reply]
Not sure if I want to go that route as I am not changing the default registration but only providing another registration route to be done by an admin or someone with permissions.

**Comment by gl2007** [comment, marked as answer]
Thanks for your answers. Here is how I solved it: 
First made the "$tables" protected as @kenjis suggested. 

Next, I added the code below to my registration method at the top. 
`        $authConfig   = config('Auth');
        $this->tables = $authConfig->tables;`

Then I removed all "user" registration based actions/redirections etc. and kept the validations/events/group registration in place. It all worked smoothly.

Thanks!
